cmd/lotus-shed/dagspliter/dagsplitter.go

```diff
@@ -122,104 +120,132 @@ func (b *builder) getSize(nd ipld.Node) (uint64, error) {
+	rn, err := mdag.NewRawNodeWPrefix(rootNode.RawData(), &cid.V1Builder{
+		Codec:    cid.Raw,
+		MhType:   p.MhType,
+		MhLength: p.MhLength,
+	})
 	if err != nil {
-		return 0, err
+		return nil, xerrors.Errorf("adding raw boxed node: %w", err)
 	}
 
-	bid := b.getBox(headSize)
-	_, _ = fmt.Fprintf(os.Stderr, level+"put %s into box %d\n", head, bid)
-
-	if headSize+b.boxes[bid].used > b.chunk { // too big for this box, need more boxes
-		p := head.Prefix()
-
-		b.boxes[bid].nodes = append(b.boxes[bid].nodes, &BoxedNode{
-			Prefix: Prefix{
-				Version:  p.Version,
-				Codec:    p.Codec,
-				MhType:   p.MhType,
-				MhLength: int64(p.MhLength),
-			},
-			Raw: cid.NewCidV1(cid.Raw, head.Hash()),
-		})
-
-		{
-			// TODO: wasn't the ipfs blockstore supposed to operate on multihashes?
-			rn, err := mdag.NewRawNodeWPrefix(headNd.RawData(), &cid.V1Builder{
-				Codec:    cid.Raw,
-				MhType:   p.MhType,
-				MhLength: p.MhLength,
-			})
-			if err != nil {
-				return 0, xerrors.Errorf("adding raw boxed node: %w", err)
-			}
+	if err := b.serv.Add(ctx, rn); err != nil {
+		return nil, xerrors.Errorf("storing raw boxed node: %w", err)
+	}
 
-			if err := b.serv.Add(ctx, rn); err != nil {
-				return 0, xerrors.Errorf("storing raw boxed node: %w", err)
-			}
-		}
+	return rn, nil
+}
 
-		_, _ = fmt.Fprintf(os.Stderr, level+"^ make %s boxed\n", head)
+func (b *builder) add(ctx context.Context, head cid.Cid, headNd ipld.Node, level string) error { // returns box id with head
+	// FIXME: Rename to cumulative/total size. Head should track just the top node size.
+	headSize, err := b.getTotalSize(headNd)
+	if err != nil {
+		return xerrors.Errorf("getting node total size: %w", err)
+	}
 
-		for l, subnd := range headNd.Links() { // todo maybe iterate in reverse
-			subNd, err := b.serv.Get(ctx, subnd.Cid)
-			if err != nil {
-				return 0, xerrors.Errorf("getting subnode: %w", err)
-			}
+	_, _ = fmt.Fprintf(os.Stderr, level+"put %s into box %d\n", head, b.boxID())
 
-			subSize, err := b.getSize(subNd)
-			if err != nil {
-				return 0, xerrors.Errorf("getting subnode size: %w", err)
-			}
+	if b.fits(headSize) {
+		_, _ = fmt.Fprintf(os.Stderr, level+"^ make internal, %dmib\n", (headSize+b.box().used)/mib)
+		b.packNodeCid(head, headSize)
+		return nil
+	}
 
-			if subSize+b.boxes[bid].used > b.chunk { // sub node too big
-				subBox, err := b.add(ctx, subnd.Cid, subNd, level+" ")
-				if err != nil {
-					return 0, xerrors.Errorf("processing subnode: %w", err)
-				}
+	// Too big for the current box. We need to examine its links to see
+	// how to partition it.
+
+	// First the parent node.
+	// FIXME: How to check the size of the parent node without taking into
+	//  account the children? The Node interface doesn't seem to account for
+	//  that so we are going directly to the Block interface for now.
+	//  (If this is indeed the correct way reuse the raw data for the extract
+	//   call next.)
```

Comment:
That's a bit annoying, and something that could be better about dag-pb, but it's likely not changing any time soon.. maybe with unixfs v2 we can get node count / total raw byte size entries.

For now we can probably get away with not accounting non-file data well, and just have some % overhead when accounting space (obviously that will break horribly with small files, but it should be good enough in the average case)